lib/internal/modules/esm/package_config.js

```diff
@@ -1,69 +1,64 @@
 
-const {
-  StringPrototypeEndsWith,
-} = primordials;
-const { URL, fileURLToPath } = require('internal/url');
-const packageJsonReader = require('internal/modules/package_json_reader');
-
-/**
- * @typedef {object} PackageConfig
- * @property {string} pjsonPath - The path to the package.json file.
- * @property {boolean} exists - Whether the package.json file exists.
- * @property {'none' | 'commonjs' | 'module'} type - The type of the package.
- * @property {string} [name] - The name of the package.
- * @property {string} [main] - The main entry point of the package.
- * @property {PackageTarget} [exports] - The exports configuration of the package.
- * @property {Record<string, string | Record<string, string>>} [imports] - The imports configuration of the package.
- */
-/**
- * @typedef {string | string[] | Record<string, string | Record<string, string>>} PackageTarget
- */
+const { ArrayIsArray } = primordials;
+const { toNamespacedPath } = require('path');
+const modulesBinding = internalBinding('modules');
 
 /**
  * Returns the package configuration for the given resolved URL.
  * @param {URL | string} resolved - The resolved URL.
- * @returns {PackageConfig} - The package configuration.
+ * @returns {import('typings/internalBinding/modules').PackageConfig} - The package configuration.
  */
 function getPackageScopeConfig(resolved) {
-  let packageJSONUrl = new URL('./package.json', resolved);
-  while (true) {
-    const packageJSONPath = packageJSONUrl.pathname;
-    if (StringPrototypeEndsWith(packageJSONPath, 'node_modules/package.json')) {
-      break;
-    }
-    const packageConfig = packageJsonReader.read(fileURLToPath(packageJSONUrl), {
-      __proto__: null,
-      specifier: resolved,
-      isESM: true,
-    });
-    if (packageConfig.exists) {
-      return packageConfig;
-    }
-
-    const lastPackageJSONUrl = packageJSONUrl;
-    packageJSONUrl = new URL('../package.json', packageJSONUrl);
-
-    // Terminates at root where ../package.json equals ../../package.json
-    // (can't just check "/package.json" for Windows support).
-    if (packageJSONUrl.pathname === lastPackageJSONUrl.pathname) {
-      break;
-    }
-  }
-  const packageJSONPath = fileURLToPath(packageJSONUrl);
-  return {
+  const packageScopeConfig = modulesBinding.getPackageScopeConfig(toNamespacedPath(`${resolved}`));
+  const response = {
     __proto__: null,
-    pjsonPath: packageJSONPath,
+    pjsonPath: undefined,
     exists: false,
     main: undefined,
     name: undefined,
     type: 'none',
     exports: undefined,
     imports: undefined,
   };
+  if (ArrayIsArray(packageScopeConfig)) {
+    const {
+      0: name,
+      1: main,
+      2: type,
+      3: imports,
+      4: exports,
+      5: _manifest, // eslint-disable-line no-unused-vars
+      6: filePath,
+    } = packageScopeConfig;
+    response.name = name;
+    response.main = main;
+    response.type = type;
+    response.imports = imports;
+    response.exports = exports;
```

Comment:
Here these are potentially unparsed JSON, right? I'm guessing you didn't immediately parse them in case they're never read. So, I'm thinking it would be better to make `response.imports` and `response.exports` getters that overwrite themselves with the parse result. This would better adhere to the principle of least knowledge (a consumer doesn't need to know these need to be manually parsed) and DRYs the parsing et al.

On a related note, I think `getPackageScopeConfig` could get called multiple times for the same input. Would it be worthwhile to memoise it?

Reply:
>  So, I'm thinking it would be better to make response.imports and response.exports getters that overwrite themselves with the parse result.

That looks like a good idea! Can you make code recommendations?

> On a related note, I think getPackageScopeConfig could get called multiple times for the same input. Would it be worthwhile to memoise it?

`getPackageScopeConfig` is memoised in C++ side, but the construction of the object isn't. I'll add a TODO and follow up with it when all the tests pass.

Reply:
It would look something like this:

```js
const response = {
  name,
  // …
  get exports() {
    const value = JSONParse(exports);
    Object.defineProperty(this, 'exports', { value });
    return value;
  },
  get imports() {…},
};
```

https://jsfiddle.net/JakobJingleheimer/du2ecbLg/